sr_robot_lib/src/MST.cpp

```diff
@@ -11,130 +11,249 @@
+  * @brief This is a class for accessing the data from the MST tactiles.
+  */
 
 #include "sr_robot_lib/MST.hpp"
 #include <sr_utilities/sr_math_utils.hpp>
 
 #include <memory>
 #include <string>
 #include <vector>
+#include <ros/console.h>
 
-#define NUMBER_OF_SENSORS 7
+/// The number of sensitive taxels on each MST tactile sensor
+#define NUMBER_OF_TAXELS 17
+#define MAGNETIC_DATA_BYTES       77
+#define TEMPERATURE_DATA_BYTES    27
 
 namespace tactiles
 {
-  template<class StatusType, class CommandType>
-  MST<StatusType, CommandType>::MST(ros::NodeHandle nh, std::string device_id,
-                                      std::vector<generic_updater::UpdateConfig> update_configs_vector,
-                                      operation_mode::device_update_state::DeviceUpdateState update_state,
-                                      boost::shared_ptr<std::vector<GenericTactileData> > init_tactiles_vector)
-          : GenericTactiles<StatusType, CommandType>(nh, device_id, update_configs_vector, update_state)
+/**
+  * MST object constructor.
+  *
+  * @param nh ROS node handle.
+  * @param device_id The device id.
+  * @param update_configs_vector A vector containing information regarding which type of data to poll.
+  * @param update_state Defines current operation state of the sensor (INITIALIZATION or OPERATION).
+  * @param init_tactiles_vector A generic tactile data vector containing sensor information (e.g. Serial Number).
+  */
+template<class StatusType, class CommandType>
+MST<StatusType, CommandType>::MST(ros::NodeHandle nh, std::string device_id,
+                                    std::vector<generic_updater::UpdateConfig> update_configs_vector,
+                                    operation_mode::device_update_state::DeviceUpdateState update_state,
+                                    boost::shared_ptr<std::vector<GenericTactileData> > init_tactiles_vector) :
+  GenericTactiles<StatusType, CommandType>(nh, device_id, update_configs_vector, update_state)
+{
+  diagnostic_data = *init_tactiles_vector;
+  initialise_tactile_data_structure(init_tactiles_vector);
+}
+
+/**
+  * Initialise relevant MST sensor data arrays.
+  */
+template<class StatusType, class CommandType>
+void MST<StatusType, CommandType>::initialise_tactile_data_structure(
+                                          boost::shared_ptr<std::vector<GenericTactileData>> init_tactiles_vector)
+{
+  // Initialise the vector of tactiles
+  this->all_tactile_data =
+                boost::shared_ptr<std::vector<AllTactileData>>(new std::vector<AllTactileData>(this->nb_tactiles));
+
+  for (uint8_t id_sensor; id_sensor < this->nb_tactiles; ++id_sensor)
   {
-    diagnostic_data = *init_tactiles_vector;
-    publisher = std::make_shared<ros::Publisher>(nh.advertise<sr_robot_msgs::MSTPalm>("mst", 1));
+    // For each sensor found, add the type and diagnostics data (freq, sw version, git version, status check)
+    MSTData mst_diagnostics(init_tactiles_vector->at(id_sensor));
```

Comment:
Is this intermediate local variable `mst_diagnostics` needed?
We seem to be copying the data two times for no reason. (It doesn't seem to make the code more readable either).

Reply:
Ok so, currently, for every sensor type, when we initialize all_tactile_data (which is later used by each specific tactile sensor publisher) we also always provide the up-to-date diagnostics data read from these sensors (see [here](https://github.com/shadow-robot/sr_core/pull/232/files) for PSTs, UBI0, Biotac). I'm also struggling to understand why this is actually necessary (if at all)

Reply:
The question now is, shall I do things differently for the MSTs? Or should we refactor this for all sensors? (the latter being out of the scope of this story)

Reply:
Sorry, I wasn't clear. what I mean is why do we do:

```
MSTData mst_diagnostics(init_tactiles_vector->at(id_sensor));
this->all_tactile_data->at(id_sensor).mst = mst_diagnostics;
```
instead of simply doing:
```
this->all_tactile_data->at(id_sensor).mst = init_tactiles_vector->at(id_sensor);
```

Why using an intermediate variable?

Reply:
That is due to this: https://github.com/shadow-robot/sr_core/blob/af065d33caa1b6b8ed9e1bfe2fdbc1e481b2b4d8/sr_hardware_interface/include/sr_hardware_interface/tactile_sensors.hpp#L434-L438
If I understand correctly, the changes you're suggesting would require me to change `AllTactileData` to something like:
```
struct AllTactileData
{
  std::string type;
  BiotacData biotac;
  PST3Data pst;
  GenericTactileData mst;
  UBI0Data ubi0;
};
```
Where only the MST's would not have a designated class of their own - which I agree, for now, it may not be necessary, but in the future, we might want to add more parameters to its constructor. But I would also argue that, if we are to change this, we could even check we are even required to provide the diagnostics data at all

Reply:
Actually, now that I take a second look, this change would not even be possible, as MST needs to be of type `MSTData` in order for me to have access to `magnetic_data`, `temperature_data` and `status_check`. See bellow:
 ```
 this->all_tactile_data->at(id_sensor).mst.magnetic_data = sensor_data_.tactiles[id_sensor].magnetic_data;
 this->all_tactile_data->at(id_sensor).mst.temperature_data = sensor_data_.tactiles[id_sensor].temperature_data;
 this->all_tactile_data->at(id_sensor).mst.status_check = sensor_data_.tactiles[id_sensor].status;
```

Reply:
Are happy to resolve this?

Reply:
OK, understood now.